Fast clone graph

Eh2406 opened the pull request:
This is a redesign of the `Graph` used by the `Resolver`. The `Resolver` is usually invoked with a lockfile, so the first thing we try will work, and any work we to to prepare for backtracking is wasted. A plane `Graph` will have an `O(n)` clone, way too expensive. An `Rc<Graph>` has `O(1)` clone but is no savings as we just end up with `O(n)`  make_mut. Before the PR we have a `Graph` built from im-rs parts, this gives use `O(1)` clone at only `O(ln(n))` overhead on every operation. This PRs `StackGraph` has `O(ln(number of clones))` clone and no overhead! There is a catch, if we have to backtrack then we do a `O(delta)` reset. In effect this PR moves work from the happy "lockfile" path to the "backtracking" path.

rust-highfive commented:
r? @ehuss

(rust_highfive has picked a reviewer for you, use r? to override)

bors commented:
:umbrella: The latest upstream changes (presumably #7174) made this pull request unmergeable. Please resolve the merge conflicts.

Eh2406 commented:
Rebased.

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -1,52 +1,227 @@
+/// This is a directed Graph structure. Each edge can have an `E` associated with it,
+/// but may have more then one or none. Furthermore, it is designed to be "append only" so that
+/// it can be queried as it would have bean when it was smaller. This allows a `reset_to` method
+/// that efficiently undoes the most reason modifications.
```
Could the comments here and the fields below be expanded to indicate what all the integer pointers are intended for? From reading this I don't have a great grasp of how this is implemented, and it looks pretty complicated internally. To clarify I don't doubt it, was just hoping to not have to read so closely below after reading this :)

Eh2406 replied:
I will add a new type for each use of `usize` and some comments.

A fair bit of the complexity comes from the `or none` support. This is needed as a lockfile knows all the packages eth package depends on but does not know the real `Dependency` objects. This data structure supports interweaving of adding links with edge data and links that do not have data. We use both but we don't in practice interweave them. I wonder if there is a way to simplify the structure if we ban interweaving...

alexcrichton commented:
I'll try to get to this tomorrow morning, it looks like it's going to be pretty deep into data structure land so I'm not quite ready to do that today!

Eh2406 commented:
Rebased, simplified, and added some comments. What else can be made clearer?

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -1,52 +1,219 @@
+            .nodes
+            .entry(node.clone())
+            .or_insert_with(IndexMap::new)
+            .entry(child.clone())
```
I think this and the previous clone can be removed?

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -1,52 +1,219 @@
+                entry.insert(edge_index);
+            }
+            Entry::Occupied(_) => {
+                // this pare is already linked
```
s/pare/pair/

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -1,52 +1,219 @@
+
+    /// connect `node`to `child` associating it with `edge`.
+    /// Note that if this and `link` are used on the same graph
+    ///      odd things may happen when `reset_to` is called.
```
This seems like a bit of an odd restriction which may be a holdover from some old version of the resolver, would it be possible to only have either `link` or `add_edge`?

Eh2406 replied:
`link` is only needed as a lockfile knows all the packages each package depends on but does not know the real `Dependency` objects ([hear](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/resolver/encode.rs#L265)). This could be a `Graph<_, ()>` but it needs to go into a `Resolve` so it needs to be  `Graph<_, Dependency>`. I am open to suggestions for how to work around this!

Having had a chance to sleep on it, I don't remember why I thought it would be a problem. I will just remove the "Note".

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -1,52 +1,219 @@
+                            next: None,
+                            previous: Some(edge_index),
+                        });
+                        let new_index: NonZeroEdgeIndex =
```
Is there much gained here from using `NonZero*` instead of just `usize`? I find the `NonZero*` stuff here to be somewhat wordy and detracting from how it's just a pointer into the array

Eh2406 replied:
8 bytes per edge. Not worth the boilerplate. Removed.

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -154,17 +324,251 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
+///  - no overhead over using the `Graph` directly when modifying
+/// Is this too good to be true? There are two caveats:
+///  - It can only be modified using a strict "Stack Discipline", only modifying the biggest clone
+///    of the graph.
```
This seems like an interesting restriction, but one you could subvert as well, right? Is this what the asserts are intended for in the `borrow` function to dynamically assert this invariant?

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -1,52 +1,219 @@
 
-    pub fn edges(&self, from: &N) -> impl Iterator<Item = &(N, E)> {
-        self.nodes.get(from).into_iter().flat_map(|x| x.iter())
+    pub fn edges(&self, from: &N) -> impl Iterator<Item = (&N, Edges<'_, E>)> {
```
This seems somewhat odd, but the idea here is that you can implicitly associate a list of data with an edge between nodes? I guess this was previously handled where `E` was `Vec<T>` and nowadays `E` is just `T`, right?

Eh2406 replied:
Yes. The old Graph can only have one `E` per link. So `E` was `Rc<Vec<Dependency>>`. Most of the time there is only 1 Dependency; the 2 cache misses did not by us mutch. The new Graph can store multiple `E` per link, so it needs to provide an iterator. The disadvantages of storing them in a linked list only come on the rare occasion that we have more than one Dependency. The other advantage is that `reset_to` is not possible (at zero cost) if we hand out `&mut E`.

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -154,17 +324,251 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
+    type Item = &'a E;
+
+    fn next(&mut self) -> Option<&'a E> {
+        while let Some(edge_link) = self.index.and_then(|old_index| self.graph.get(old_index)) {
```
Is the `get` here required to be a fallible lookup? I'd imagine that if we're iterating all internal edge pointers should already be valid

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -154,17 +324,251 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
+                .nodes
+                .get(from)
+                .and_then(|x| x.get(to).copied())
+                .filter(|idx| idx < &self.age.len_edges),
```
The extra filtering done here in the view is somewhat odd, can you be sure to add comments (not sure where, just somewhere would be fine) as to why it's necessary?

Eh2406 commented:
Thank you for the on-point feedback. I added some comments and responded to some of your points. I did not yet have time for the multi-paragraph-long description. I will need to think through how to describe only the complexity required to read this module vs providing enough context to grok why it is what we need. You other points should become clearer if (when?) I describe the hole picture well.

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -154,17 +333,271 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
+}
+
+/// A RAII gard that allows getting `&` references to the prefix of a `Graph` as stored in a `StackGraph`.
+/// Other views of the inner `Graph` may have added things after this `StackGraph` was created.
```
Hm I'm not sure how this sentence is true, once this is created it freezes `Graph` inside it because of the borrow on the `RefCell`, right?

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -154,17 +333,271 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
+    fn next(&mut self) -> Option<&'a E> {
+        while let Some(edge_link) = self.index.and_then(|idx| {
+            // Check that the `idx` points to something in `self.graph`. It may not if we are
+            // looking at a smaller prefix of a larger graph.
```
I think this is something I don't quite understand, I thought the point of `reset_to` was that these sort of extraneous edges were pruned out? How do they linger around?

Eh2406 commented:
I wrote several paragraphs of words, I can no longer tell if they mean anything. Let me know if any of it makes sense and is worth having around. I added a test that I hope helps show the queries this supports.

Having spent so long trying to describe what I meant by "Stack Discipline", I am now wondering if it would be possible to detect when a deep clone is needed automatically. At runtime given how the Resolver uses the `StackGraph`, the ancere is never, but it may be easier to have code to deep clone when needed then it is to describe how to arrange not to need deep clones. I will continue to think about it. Now that I have started to think about that, if we conservatively always to a deep clone before befor `reset_to` we still get the same BigO. I will give that a try, "Stack Discipline" may be premature optimization. I will try.

bors commented:
:umbrella: The latest upstream changes (presumably #7186) made this pull request unmergeable. Please resolve the merge conflicts.

Eh2406 commented:
I am now setup on my new laptop, so I can do more accurate bench-marking. The version that does a deep clone any time it may be needed is slower then master when backtracking is involved. More experiments are needed.

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -1,52 +1,284 @@
+//! `dependency_graph_so_far.clone();`. To make this more annoying the first thing we try will
+//! probably work, and any work we do to prepare for the next iteration is wasted. If we had a
+//! `undo_activate` we could be much more efficient, completely remove the `.clone()` and just
+//! `undo_activate` if things tern out to not work. Unfortunately, making shore `undo_activate`
```
s/shore/sure/

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -1,52 +1,284 @@
 
+/// This is a directed Graph structure. Each edge can have an `E` associated with it,
+/// but may have more then one or none. Furthermore, it is designed to be "append only" so that
+/// it can be queried as it would have bean when it was smaller. This allows a `reset_to` method
```
s/bean/been/

Eh2406 commented:
Added code to determine when a deep clone is need. This cuts into the wins from this PR, but means that it does not need "Stack Discipline" to be correct, and is faster or the same on all the things I have tried.

Over all, this has been a lot of work to beat `im-rs` for a small win. @bodil I am very impressed.

alexcrichton commented:
Hm I may be getting lost again. I'm sort of confused, if this is all a pretty small win should we stick with the im-rs data structures? Those are presumably much easier to read/write, but especially with the new `Clone` implementation this is much more complex to understand.

Eh2406 commented:
3 weeks ago when I started committing to this branch I was definitely planning to say: "this adds a large amount of complexity, is it worth it for the performance boost?" I seem to have forgotten to make that clere. Sorry! It is entirely reasonable for us to close this becust the wins are not worth the code. But let's get quantitative:

% improve by commit and commend
-----

Etch average of 3 runs on my new laptop as idle as I can get it, and with `-Zno-index-update`.
Please report results on your workload.

[2000 crate stress test](https://github.com/rust-lang/cargo/pull/6967#issuecomment-499883732)

commit | update | generate-lockfile | check
---------|---------|--------------------|---
ab1d533a7265b237ff2c0e8e9fae67d70211f1ca |  13.29% | 12.73% | 2.58%
5aff94507e2f86ee912b0c3c08f9fb2faa9e2e6a |  -11.16% | -11.34% | 3.86%
518028b60fca16e66e7639b562a8d0b7ab2deedf (head) | 1.85% | 10.32% | 3.18%

on cargo's repo

commit | update | generate-lockfile | check
---------|---------|--------------------|---
ab1d533a7265b237ff2c0e8e9fae67d70211f1ca |  5.75% | -0.77% | 3.13%
5aff94507e2f86ee912b0c3c08f9fb2faa9e2e6a |  0.35% | 3.27% | 3.29%
518028b60fca16e66e7639b562a8d0b7ab2deedf (head) | 3.74% | 3.32% | 4.58%

Is ~5% on no-op checks of Cargo big enough to justify all this new code?

bors commented:
:umbrella: The latest upstream changes (presumably #7191) made this pull request unmergeable. Please resolve the merge conflicts.

Eh2406 commented:
A profile of no-op `cargo check` on cargo spends 2/3 of the runtime of main in `cargo::sources::registry::index::RegistryIndex::query_inner_with_online` (cc #6908). So any measurable difference from changes to the Resolver is impressive.

Eh2406 commented:
Rebased

Eh2406 commented:
> I still find the implementation to be somewhat difficult to follow and can't figure out how to walk away with a feeling of certainty of its correctness.

That is strong evidence that the additional complexity of this PR is not worth it. Getting the borrow checker happy with this took way more code then I was expecting when I started out. 

> Coming back to the measurements, though, it looks like this is a max 5% win in the resolution phase of Cargo, right? If so, that seems somewhat small for the weight of the PR itself, especially when the 5% doesn't translate well to the compilation backend which also has lots of syscalls and such which can be slow. 

Not precisely. It is a 5% win of a full end-to-end no-op build of cargo. `cargo> time cargo check`. This is impressive as the resolver (except for reading the index) is already a pretty small part of the runtime. I can make more impressive numbers by benchmarking the Resolver by itself, or by micro-benchmarking the Graph by itself. You're generally more concerned with the "happy lockfie path" thru the resolver. So I was excited to have a end-to-end improvement to show you. With all the work we have done making the resolver faster in other ways, this is only a 5% win. Law of diminishing returns and all... 

> Are you thinking though that future features of Cargo would rely on this performance improvement more so than we do today?

I don't see us getting more bottlenecked on Graphs performance, so No. On the other hand the approach of making clones faster (by keeping the structure sorted by insertion and only cloning the length) could also be applied to all the other components of `BacktrackFrame`/`Context`, so Yes I can see this technique being tried again. Especially as we see the overhead of Public Private Dependencies on real workloads, or #6908 moves more of the runtime back to the resolver. The Resolver works hard to avoid `BacktrackFrame::clone`s. If it gets cheap enough then there are several opportunities for simplifying the code (#6967 for a start). Complexity in the Structs verse in the Algorithm...

alexcrichton commented:
Oh sorry I should also clarify that my inability to wrap my head around this is probably also related to me, at some point it's probably good to get a second opinion.

Mind getting some numbers for pure resolver performance? I agree 5% overall is quite compelling, and I suspect the resolver numbers will be even better!

I personally like the idea of simplifying the resolver as much as possible internally if we can, and I wouldn't consider the data structures part of that complexity since they're typically very well encapsulated

Eh2406 commented:
I added a

 ```rust
impl Drop for ResolverProgress {
    fn drop(&mut self) {
        println!("{}", (self.start.elapsed() - self.deps_time).as_secs_f32());
    }
}
```
in [resolver/types](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/resolver/types.rs). This should report the time in the resolver without the time in activations. (Witch is mostly spent opening index related files.) Re-doing the same profiling using this time. Etch average of 15 runs on my new laptop, and with `-Zno-index-update`. Note that `check` drops 2 Resolvers (I don't know why) so I added the results together for each run.

repo| update | generate-lockfile | check
---------|---------|--------------------|---
[2000 crate stress test](https://github.com/rust-lang/cargo/pull/6967#issuecomment-499883732)|  15.41% | 14.97% | 34.31%
cargo's repo |  17.95% | 19.43% | 18.01%

This is % improvement dc914ba9170c89344e039a7fc7ae5ffbbc641e5f over 4bdb6fd371111e2ec80811822c130dde66af18df. Bigger numbers are better.

nrc commented on src/cargo/util/graph.rs:
```diff
@@ -1,52 +1,284 @@
-    pub fn link(&mut self, node: N, child: N) -> &mut E {
-        self.nodes
+    /// connect `node`to `child` with out associating any data.
+    pub fn link(&mut self, node: N, child: N) {
```
Not clear to me if link means create an edge or something else. I think in general I'm just a bit confused between links and edges

nrc commented on src/cargo/util/graph.rs:
```diff
@@ -154,17 +389,333 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
+///  - the clone has no overhead to read the `Graph` as it was
+///  - no overhead over using the `Graph` directly when modifying the biggest clone
+/// Is this too good to be true?
+///  - If a modification (`&mut` method) is done to a smaller older clone then a full `O(N)`
```
stm this is a bit of a performance footgun. I think it would be better for such a modification to fail and have an explicit deep clone method.

nrc commented on src/cargo/util/graph.rs:
```diff
@@ -154,17 +389,333 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
+    /// needed, and runs `reset_to` to get the `Graph` into the state it was in.
+    ///
+    /// It is the responsibility of the caller to set `self.age` and call `add_to_other_refs` after the new age is determined.
+    fn activate(&mut self) -> RefMut<'_, Graph<N, E>> {
```
Why is this called `activate`? And could the comment explain why the caller has to maintain the object invariants.

nrc commented on src/cargo/util/graph.rs:
```diff
@@ -154,17 +389,333 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
+    /// The shared list of all extant references to the same `Graph`.
+    /// The largest one is allowed to reset and expend the `Graph` without doing a deep clone.
+    /// There can be more then one clone at the same age, so each age is associated with a count.
+    other_refs: Rc<RefCell<BTreeMap<GraphAge, usize>>>,
```
I can't really explain why, but this seems like a sub-optimal way to track if we're the largest graph. I'd expect you could have something less complex or more intrinsic?

bors commented:
:umbrella: The latest upstream changes (presumably #7216) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #7425) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #7452) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #7361) made this pull request unmergeable. Please resolve the merge conflicts.

Eh2406 commented:
This is not going to get the attention that is needed to merge. And at this point I'd rather see the effort go into PubGrub not into optimizing this resolver. 

A lot of ink has gone into it, a more responsible maintainer would turn it into documentation before closing. But I am not that responsible.
